### PR TITLE
[runtime] Use explicit protected regions to handle more EH edge cases

### PR DESCRIPTION
**Pull request by alexanderkyte**
Some of the edge cases for llvmonly exception handling were only exposed by the mono/tests tests, not the bcl tests. It seems that a few things were broken. Throwing in a protected region's clauses could break when there was nesting. Finally clauses also didn't necessarily chain with nesting. This is a fix to both, also bringing with it an explicit interval tree (a bit of a hack at one, due to the queries we're making) in order to make this kind of thing a lot easier to write than the offset manipulation of the previous implementation. This fixes crashes or asserts in exception7, exception4, and a lot of other exception tests. Not everything is perfect, but this implementation started by mimicking the old control flow, so there shouldn't be regressions. 

Let's see what CI thinks of this. This will only apply to the mobile_static bitcode lane, other failures should be outside the scope of my changes.

### CODE REVIEW

**Comment by alexanderkyte** [comment]
build bitcode

**Comment by vargaz** [comment]
Which tests fail without this ?

**Comment by alexanderkyte** [comment]
A lot of the mono/tests tests have the form exceptionN.cs . Exception4.cs was the one that exhibited the lack of nested finally execution.

**Comment by vargaz** [comment]
Those tests work for on on osx with llvmonly. Is this a linux only issue ?

File: mono/mini/mini.h [hunk 1]
```diff
@@ -1181,12 +1181,20 @@ typedef struct {
...
+
 	/*
 	 * If the current exception is not a subclass of Exception,
 	 * the original exception.
 	 */
 	guint32 thrown_non_exc;
 
+	/* 
+	 * A temporary reference to the ips for the whole
+	 * stack t			race when thrown. These are truncated when the
```
Comment (evincarofautumn) [inline mono/mini/mini.h]: Extra tabs

File: mono/mini/mini-exceptions.c [hunk 2]
```diff
@@ -255,6 +254,29 @@ build_stack_trace (struct _Unwind_Context *frame_ctx, void *state)
...
+
+static _Unwind_Reason_Code
+count_frames (struct _Unwind_Context *frame_ctx, void *state)
+{
+	uintptr_t ip = _Unwind_GetIP (frame_ctx);
+	gboolean add_ptr = add_ptr_to_stack_trace (ip);
+
+	if (add_ptr) {
+		size_t *frame_size = (size_t *)state;
+		*frame_size = *frame_size + 1;
```
Comment (evincarofautumn) [inline mono/mini/mini-exceptions.c]: `++*frame_size;`

File: mono/mini/mini-exceptions.c [hunk 3]
```diff
@@ -240,11 +240,10 @@ static gboolean show_native_addresses = TRUE;
 static gboolean show_native_addresses = FALSE;
 #endif
 
-static _Unwind_Reason_Code
-build_stack_trace (struct _Unwind_Context *frame_ctx, void *state)
+static gboolean
+add_ptr_to_stack_trace (uintptr_t ip)
```
Comment (evincarofautumn) [inline mono/mini/mini-exceptions.c]: `uintptr_t` because `_Unwind_GetIP` returns `uint64` or `uint32` depending on platform?

File: mono/mini/mini-exceptions.c [hunk 4]
```diff
@@ -3095,6 +3107,53 @@ mono_llvm_match_exception (MonoJitInfo *jinfo, guint32 region_start, guint32 reg
...
+
+		// Build stacktrace backwards, move cursor to end
+		GList *backward = forward;
+		int size = 1;
+		while(backward->next) {
+			backward = backward->next;
+			size++;
+		}
+		/* The list contains ip-gshared info pairs */
+		size = size * 2;
```
Comment (evincarofautumn) [inline mono/mini/mini-exceptions.c]: This sorta changes the meaning of `size` from “number of pairs” to “number of elements”. Either introduce a different variable or use consistent counting throughout.

File: mono/mini/mini-exceptions.c [hunk 5]
```diff
@@ -3095,6 +3107,53 @@ mono_llvm_match_exception (MonoJitInfo *jinfo, guint32 region_start, guint32 reg
...
+		mono_error_init (&error);
+		MonoArray *trace_arr = mono_array_new_checked (mono_domain_get (), mono_defaults.int_class, size, &error);
+		mono_error_assert_ok (&error);
+
+		for(int index = 0; index < size; index += 2, backward = backward->prev) {
+			g_assert(backward != NULL);
+
+			/* The list contains ip-gshared info pairs */
+			mono_array_set (trace_arr, gpointer, index, backward->data);
+			mono_array_set (trace_arr, gpointer, index + 1, NULL);
```
Comment (evincarofautumn) [inline mono/mini/mini-exceptions.c]: This is redundant if you have a new array, but might be better to leave in (or leave a comment) for clarity.

File: mono/mini/mini-llvm.c [hunk 6]
```diff
@@ -3946,70 +3937,90 @@ mono_protected_region_insert (EmitContext *ctx, MonoProtectedRegion **out, MonoE
...
 		mono_protected_region_add (ctx, val, input);
 
 		return TRUE;
-	} else if (CLAUSE_START (input) > CLAUSE_START (ex) || CLAUSE_END(ex) != CLAUSE_END (input)) {
-		// If the element is a child, and not at this level
-		for (GSList *curr = val->subintervals; curr; curr = curr->next) {
-			MonoExceptionClause *cast = mono_protected_region_elem (curr->data);
-			if (CLAUSE_START (input) <= CLAUSE_START (cast) || CLAUSE_END(cast) >= CLAUSE_END (input))
-				return mono_protected_region_insert (ctx, (MonoProtectedRegion **)&curr->data, input);
+	} else if (CLAUSE_START (input) >= CLAUSE_START (ex) || CLAUSE_END(ex) != CLAUSE_END (input)) {
```
Comment (evincarofautumn) [inline mono/mini/mini-llvm.c]: Since you always return, none of these needs to be an `else`, saving some nesting; ditto the returns and break in this branch.

File: mono/mini/mini.h [hunk 7]
```diff
@@ -2851,8 +2851,8 @@ MONO_API gboolean mono_exception_walk_trace     (MonoException *ex, MonoExceptio
 void mono_restore_context                       (MonoContext *ctx);
 guint8* mono_jinfo_get_unwind_info              (MonoJitInfo *ji, guint32 *unwind_info_len);
 int  mono_jinfo_get_epilog_size                 (MonoJitInfo *ji);
-void     mono_llvm_rethrow_exception            (MonoObject *ex);
-void     mono_llvm_throw_exception              (MonoObject *ex);
+void     mono_llvm_rethrow_exception            (MonoObject *ex, gboolean actually_throw);
```
Comment (evincarofautumn) [inline mono/mini/mini.h]: What does `actually_throw` mean?

**Comment by evincarofautumn** [comment]
Did you mean to commit the tags file?

File: mono/mini/mini-exceptions.c (hunk 3 above)
Comment (alexanderkyte) [inline mono/mini/mini-exceptions.c]: It's because the value passed to this is a _Unwind_Ptr, which is just a typedef around uintptr_t.  http://clang.llvm.org/doxygen/unwind_8h.html#ab566a2905df286d9520b814e80f2c23f

I went with using the naked type because we want to treat them as char-stars later. I can change this if it's preferred to use another type stylistically, but this seemed the easiest to read and understand what we're actually getting.

**Comment by alexanderkyte** [comment]
@evincarofautumn though github doesn't seem to show it, my commits should have addressed those notes

**Comment by luhenry** [comment]
@vargaz @alexanderkyte should we pursue with this PR or close it? How much work would it be to make it up to date?
